### ISSUE TITLE
Signal app share sheet crash

### ISSUE THREAD

**Issue by UtopianCitizen** [signalapp/Signal-iOS#4220]
- [X] I have searched open and closed issues for duplicates
- [X] I am submitting a bug report for existing functionality that does not work as intended
- [X] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
When I try to share links or files using the share sheet and clicking the signal app, it opens for a brief moment and then crashes.

### Steps to reproduce
- open twitter app
- click share button
- click share this tweet via
- select Signal app

**Actual result:** signal opens for a brief moment and then crashes 

**Expected result:**  I should be able to select the contact to send the link

### Screenshots

![66C1C93B-91F9-4E92-9966-A2299EEE3706](https://user-images.githubusercontent.com/52544653/60682212-77795480-9e60-11e9-8395-378b6e074f25.png)
![3B9A8C5E-AE90-4DA1-A243-962B7D1EF25F](https://user-images.githubusercontent.com/52544653/60682213-77795480-9e60-11e9-8117-6ac6c2246b2d.png)
![9B8117D4-C975-4DB3-BF97-321EA755B186](https://user-images.githubusercontent.com/52544653/60682214-77795480-9e60-11e9-935b-59adb61a6ee3.png)
![BA359B34-58FF-4A7D-8CD8-67F5383AFDA5](https://user-images.githubusercontent.com/52544653/60682215-77795480-9e60-11e9-9e01-32275da0e8ad.png)
![6782F957-A04F-401E-A5D0-B27372B06E25](https://user-images.githubusercontent.com/52544653/60682217-7811eb00-9e60-11e9-977a-58315e7dbf6f.png)

### Device info

**Device**: iPhone XR

**iOS version**: 12.3.1

**Signal version:** 2.40.1.3

**Comment by akhachane** [comment]
Same issue but with photos on Xs

**Comment by dlobraico** [comment]
I’ve seen this issue on an iPhone X as well.

**Comment by michaelkirk-signal** [comment]
Can you please submit a crash report?

1. On your phone, open the Settings App
2. Go to Privacy > Analytics > Analytics Data
3. Locate the crash report - you're looking for one whose title includes the word "Signal" and "Extension"
4. Select the most recent matching crash report
5. Tap the share button in the nav bar, and email the crash repot to support@signal.org with the subject "iOS Crash Log: <Github Issue Reference>"
    Example subject: iOS Crash Log: Crash on launch #111
    Example subject: iOS Crash Log: Crash when sending video #222

**Comment by dlobraico** [comment]
I’d be happy to, but I don’t see a crash report whose name includes both
Signal and Extension. There are a few that contain just Signal, but the
most recent one is from yesterday despite my having observed the crash much
more recently than that. Please advise.

**Comment by dlobraico** [comment]
Anything else I can do here?

**Comment by akhachane** [comment]
There was no crash report and I have done the most recent updates and the issue continues.  I wonder if this is an apple thing.  Though, I also have a newish phone and problem persists.

**Comment by akhachane** [comment]
So I had to get a new phone and it seems that it worked fine until I either installed or logged in to another app on the phone and the issue seems to still be there.  Not sure what it was, any thoughts?  Would this be another messaging app, or something else?

**Comment by Quorum75** [comment]
Same thing here. Not working with the official release or the beta version. It’s a real pain if you want to send something. It Has been 6 months and the problem has not been solved. 

iOS 13.4.1 
Signal 3.7.1
Signal Beta 3.7.1.0

**Comment by chuggsey** [comment]
Same...
iOS 13.4.1 on iPhone 8
Signal 3.7.1.0

**Comment by aeolusnus** [comment]
Here is what I have seen to date. There is no crash report for this, the application seems to quit, not crash. Now that there is the signal app for iPad, I was able to load there. Seems like this works there with no issue. I also think that signal worked on my iPhone either prior to installing or prior to signing in to some other apps. There must be a conflict with this app and another. Any thoughts on that being a possibility? Thanks again!

**Comment by chuggsey** [comment]
thanks for the update. I shall investigate / have a play and see if I can
track down any conflict. Don't seem to suffer the same fate on my iPhone 7
+cheers

**Comment by ccweis** [comment]
Same issue here.  Works on iPad, doesn't work on iPhone.  It makes the "Note to Self" feature far less useful.

iOS 13.5.1 on iPhone 11 Pro Max
Signal 3.10.1.1

**Comment by ardakara** [comment]
Same issue on iOS 13.6 on iPhone 11 Pro Max
Signal 3.13.2

**Comment by Quorum75** [comment]
Many people have this issue and it goes back to a year now and Signal dev team just doesn't seem to care. 😡

**Comment by dividuum** [comment]
Probably same issue: On sharing anything, Signal opens, you briefly see the blueish background and it immediately closes: Syslog shows this seems to be a memory limit issue and the process is terminated upon reaching 120MB:

```
Aug  2 14:36:36 SignalShareExtension(RunningBoardServices)[12841] <Notice>: Sending handshake request attempt #1 to server
Aug  2 14:36:36 SignalShareExtension(RunningBoardServices)[12841] <Notice>: Creating connection to com.apple.runningboard
Aug  2 14:36:36 runningboardd(RunningBoardServices)[28] <Notice>: Resolved XPC Service org.whispersystems.signal.shareextension (SignalShareExtension.appex) with host pid 12293, variant 2, scope 1, pid 12841, and euid 501
Aug  2 14:36:36 SignalShareExtension(RunningBoardServices)[12841] <Notice>: Handshake succeeded
Aug  2 14:36:36 SignalShareExtension(RunningBoardServices)[12841] <Notice>: Identity resolved as xpcservice<org.whispersystems.signal.shareextension>:12841
Aug  2 14:36:36 SignalShareExtension(PlugInKit)[12841] <Notice>: Bootstrapping; Bootstrap complete. Ready for handshake from host.
Aug  2 14:36:36 SignalShareExtension(FrontBoardServices)[12841] <Notice>: FBSWorkspace connecting to endpoint : <private>
Aug  2 14:36:36 SignalShareExtension(FrontBoardServices)[12841] <Notice>: FBSWorkspace registering source: <private>
Aug  2 14:36:36 SignalShareExtension(FrontBoardServices)[12841] <Notice>: FBSWorkspace connected to endpoint : <private>
Aug  2 14:36:36 SignalShareExtension(RunningBoardServices)[12841] <Notice>: Added observer for process assertions expiration warning: <_RBSExpirationWarningAssertion: 0x2831f2ec0; identifier: com.apple.runningboardservices.processExpirationWarningForHandle; reason: observation; valid: YES>
Aug  2 14:36:36 SignalShareExtension(FrontBoardServices)[12841] <Notice>: FBSWorkspace already connected to endpoint : <private>
Aug  2 14:36:36 SignalShareExtension(libAccessibility.dylib)[12841] <Notice>: Retrieving resting unlock: 0
Aug  2 14:36:36 SignalShareExtension(XCTTargetBootstrap)[12841] <Notice>: Registering for test daemon availability notify post.
Aug  2 14:36:36 SignalShareExtension(XCTTargetBootstrap)[12841] <Notice>: notify_get_state check indicated test daemon not ready.
Aug  2 14:36:36 SignalShareExtension(PlugInKit)[12841] <Notice>: [u XXXXXXXX] [(null)((null))] Prepare received as euid = 501, uid = 501, personaid = 1000, type = DEFAULT, name = <private>
Aug  2 14:36:36 SignalShareExtension(PlugInKit)[12841] <Notice>: [u XXXXXXXX] [<private>(<private>)] Set sole personality.
Aug  2 14:36:36 SignalShareExtension(PlugInKit)[12841] <Notice>: [u XXXXXXXX] [<private>(<private>)] Begin using received as euid = 501, uid = 501, personaid = 1000, type = DEFAULT, name = <private>
Aug  2 14:36:36 SignalShareExtension(RunningBoardServices)[12841] <Notice>: Invalidating plugin handshake assertion id XXXXXX
Aug  2 14:36:36 kernel[0] <Notice>: memorystatus: set assertion priority(9) target SignalShareExtension:12841
Aug  2 14:36:36 SignalShareExtension(CoreFoundation)[12841] <Error>: Couldn't read values in CFPrefsPlistSource<0x280ac7180> (Domain: group.org.whispersystems.signal.group, User: kCFPreferencesAnyUser, ByHost: Yes, Container: (null), Contents Need Refresh: Yes): Using kCFPreferencesAnyUser with a container is only allowed for System Containers, detaching from cfprefsd
Aug  2 14:36:36 SignalShareExtension(libsystem_containermanager.dylib)[12841] <Notice>: Requesting group container lookup; identifier = <private>, euid = <private>, uid = <private>, personaid = <private>, type = <private>, name = <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(Security)[12841] <Notice>: Adding securityd connection to pool, total now 1
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(CoreServicesInternal)[12841] <Notice>: kExcludedFromBackupXattrName set on path: <private>
Aug  2 14:36:36 SignalShareExtension(libnetwork.dylib)[12841] <Notice>: nw_path_evaluator_start [XXXXXXXXXX <NULL> generic, indefinite]
Aug  2 14:36:36 SignalShareExtension(UIKitCore)[12841] <Notice>: Received memory warning.
Aug  2 14:36:36 SignalShareExtension(UIKitCore)[12841] <Notice>: Received memory warning.
Aug  2 14:36:36 SignalShareExtension(UIKitCore)[12841] <Notice>: Received memory warning.
Aug  2 14:36:36 kernel[0] <Notice>: EXC_RESOURCE -> SignalShareExtension[12841] exceeded mem limit: InactiveHard 120 MB (fatal)
Aug  2 14:36:36 kernel[0] <Notice>: 419660.666 memorystatus: killing_specific_process pid 12841 [SignalShareExtension] (per-process-limit 9) 122880KB - memorystatus_available_pages: 38544
Aug  2 14:36:36 backboardd(IOKit)[60] <Notice>: Connection removed: IOHIDEventSystemConnection uuid:XXXXXXXX pid:12841 process:SignalShareExtension type:Passive entitlements:0x0 caller:BackBoardServices: <redacted> + 368 attributes:{
Aug  2 14:36:36 osanalyticshelper(OSAnalytics)[12222] <Notice>: Process SignalShareExtension [12841] killed by jetsam reason per-process-limit
Aug  2 14:36:36 ReportCrash[12839] <Notice>: Formulating fatal report for corpse[12841] SignalShareExtension
```

**Comment by aeolusnus** [comment]
Here are the apps on my phone which are not on my iPad. I assume one of these is creating a conflict crashing the signal app: 


1.1.1.1
--
AAA
Adobe Scan
Alarm.com
Ally Auto
Amazon
Amazon Go
Amazon Music
Amex
Arro
Arts & Culture
Asana
Auction.com
Authenticator
Auto Club
Banking
BigY
Call Filter
CamCard
Capital One
Capital One
CBS Sports
Chrome River
Circle
Citizen
Classroom
CR SNAP
Curb
eFax
Enterprise
Facebook
Family
Link
findyourzo
First Stop Health
Fitbit
Flight Tracker
FreeOTP
Google Calendar
Google Fit
Google Home
Google Maps
Google Pay
Google Photos
Hangouts
Health4Me
Hertz
Hik-Connect
iFax
Instagram
iPostaI1
iPrint&Scan
JetBIue
Libby
Linkedln
LiveSafe
LoopNet
Lyft
MarketWatch
Merlin Bird ID
Messenger
MTA
My Leviton
My MBFS
My Verizon
Nest
NYC Subway
OneDrive
OneNote
onTime
OpenTable
Orbi
OverDrive
Pandora
Pinterest
PIanGrI'd
Pocket
Press Market Rewards
Pret A Manger
Prime Video
Progressive
Ring
Ritual
Sheets
Shell
Shopping
SimplyE
SmartCam+
Smartsheet
SnoreLab
Splash Car Washes
SpotHero
Starbucks
Strava
Sweetgreen
TCC
Thrive Market
TIAA
TikTok
Todoist
Twitter
Uber
UberConference
UHC
UI
Venmo
Vivino
Waze
WebsterBank
Whole Foods
Workday
Zappos
Whatsapp

**Comment by ardakara** [comment]
Every single share from iOS share sheet still crashes, no change as of 5.0.1

**Comment by ooottttooo** [comment]
Same here. Have deleted Signal. Re-downloaded from AppStore. Still quitting when trying to share s.th. 
Any idea what I could do?

**Comment by ardakara** [comment]
Progress - I’ve solved this problem on my device by turning off access to contacts for Signal since it seemed like it was being killed for high memory use while trying to load the contacts list for share like below  

2021/01/10 17:36:05:484  💛 [AppReadiness.m:131 -[AppReadiness setAppIsReady]]: 
2021/01/10 17:36:05:485  💛 [OWSContactsManager.m:111 -[OWSContactsManager setup]_block_invoke]: loading 292 signal accounts from cache.
2021/01/10 17:36:05:572  💛 [Bench.swift:57 Bench(title:logIfLongerThan:logInProduction:block:)]: [Bench] title: AppReadiness.willBecomeReady, duration: 87.73ms
2021/01/10 17:36:05:588  💛 [StorageServiceManager.swift:300 run()]: cleanUpUnknownData
2021/01/10 17:36:05:589  💛 [AppExpiry.swift:65 logExpirationState()]: Build expires in 71 days
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Config.SupportedIsEnabled: ios.kbs = true
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:261 ]: Config.SupportedIsEnabled: ios.uuidSafetyNumbers = nil
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:261 ]: Config.SupportedIsEnabled: ios.groupsV2InviteLinksV2 = nil
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Config.SupportedIsEnabled: ios.profilesForAll = true
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Config.SupportedIsEnabled: ios.groupsV2autoMigrations = true
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:261 ]: Config.SupportedIsEnabled: ios.groupsV2manualMigrations = nil
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:261 ]: Config.SupportedIsEnabled: ios.groupsV2blockingMigrations = nil
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:261 ]: Config.SupportedIsEnabled: ios.groupCallingKillSwitch = nil
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:261 ]: Config.StickyIsEnabled: ios.uuidSafetyNumbers = nil
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Config.SupportedValues: global.groupsv2.maxGroupSize = 151
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Config.SupportedValues: global.groupsv2.groupSizeHardLimit = 1001
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Config.SupportedValues: ios.clientExpiration = []
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Config.SupportedValues: research.megaphone.1 = 1:29500,44:150000,55:150000,33:150000,49:43900,7:150000,48:150000,52:150000
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Config.StickyValues: global.groupsv2.maxGroupSize = 151
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Config.StickyValues: global.groupsv2.groupSizeHardLimit = 1001
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Flag: groupCalling = 1
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Flag: groupsV2InviteLinks = 1
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Flag: groupsV2MaxGroupSizeHardLimit = 1001
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Flag: groupsV2MaxGroupSizeRecommended = 151
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Flag: groupsV2MigrationAutoMigrations = 1
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Flag: groupsV2MigrationBlockingMigrations = 0
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Flag: groupsV2MigrationManualMigrations = 0
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Flag: kbs = 1
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Flag: profilesForAll = 1
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Flag: researchMegaphone = 0
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Flag: usernames = 0
2021/01/10 17:36:05:589  💛 [RemoteConfigManager.swift:259 ]: Flag: uuidSafetyNumbers = 0
2021/01/10 17:36:05:589  💛 [Bench.swift:57 Bench(title:logIfLongerThan:logInProduction:block:)]: [Bench] title: AppReadiness.willBecomeReady group, duration: 105.30ms
2021/01/10 17:36:05:590  💛 [ShareViewController.swift:298 ensureRootViewController()]: Presenting initial root view controller
2021/01/10 17:36:05:590  💛 [ShareViewController.swift:312 presentContentView()]: Presenting content view
2021/01/10 17:36:05:591  💛 [ShareViewController.swift:179 activate()]: running post launch block for registered user: Optional(<SignalServiceAddress phoneNumber: [ REDACTED_PHONE_NUMBER:xxx250 ], uuid: [ REDACTED_UUID:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx4F ]>)
2021/01/10 17:36:05:592  💛 [ShareViewController.swift:249 checkIsAppReady()]: localAddress: Optional(<SignalServiceAddress phoneNumber: [ REDACTED_PHONE_NUMBER:xxx250 ], uuid: [ REDACTED_UUID:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx4F ]>)
2021/01/10 17:36:05:592  💛 [AppVersion.m:146 -[AppVersion appLaunchDidComplete]]: appLaunchDidComplete
2021/01/10 17:36:05:614  💛 [ShareViewController.swift:713 loadItem(unloadedItem:)]: unloadedItem: UnloadedItem(itemProvider: <NSItemProvider: 0x280979ce0> {types = (
    "public.jpeg"
)}, itemType: SignalShareExtension.ShareViewController.(unknown context at $1046cebd4).UnloadedItem.ItemType.image)
2021/01/10 17:36:05:614  💛 [ShareViewController.swift:191 activate()]: running post launch block for registered user: Optional(<SignalServiceAddress phoneNumber: [ REDACTED_PHONE_NUMBER:xxx250 ], uuid: [ REDACTED_UUID:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx4F ]>)
2021/01/10 17:36:05:623  💛 [SystemContactsFetcher.swift:353 updateContacts(completion:isUserRequested:)]: fetching contacts
2021/01/10 17:36:05:647  💛 [JobQueue.swift:214 restartOldJobs()]: marking old `running` MessageSender JobRecords as ready: 0
2021/01/10 17:36:05:695  🧡 [NSData+Image.m:779 +[NSData(Image) imageMetadataWithImageSource:imageFormat:isAnimated:]]: Could not determine transparency of image
2021/01/10 17:36:05:719  💛 [ShareViewController.swift:541 buildAttachmentsAndPresentConversationPicker()]: showing picker with attachments: [[SignalAttachment] mimeType: image/jpeg, fileSize: 176 KB]
2021/01/10 17:36:05:804  🧡 [NSData+Image.m:779 +[NSData(Image) imageMetadataWithImageSource:imageFormat:isAnimated:]]: Could not determine transparency of image
2021/01/10 17:36:05:811  🧡 [NSData+Image.m:779 +[NSData(Image) imageMetadataWithImageSource:imageFormat:isAnimated:]]: Could not determine transparency of image
2021/01/10 17:36:05:878  💛 [RemoteConfigManager.swift:500 refreshIfReady()]: skipping due to recent fetch.
2021/01/10 17:36:05:906  💛 [TSAccountManager.swift:166 updateAccountAttributesIfNecessaryAttempt()]: Skipping; lastAppVersion: Optional("[ REDACTED_IPV4_ADDRESS:...0 ]"), currentAppVersion: [ REDACTED_IPV4_ADDRESS:...0 ].
2021/01/10 17:36:05:906  💛 [TSAccountManager.swift:110 updateAccountAttributesIfNecessary()]: Success.
2021/01/10 17:36:06:098  💛 [StorageServiceManager.swift:300 run()]: restoreOrCreate
2021/01/10 17:36:06:102  💛 [StorageService.swift:194 fetchLatestManifest(greaterThanVersion:)]: 
2021/01/10 17:36:06:102  💛 [TSNetworkManager.m:399 -[TSNetworkManager makeRequestSync:completionQueue:success:failure:]]: Making Non-UD request: { GET: v1/storage/auth }
2021/01/10 17:36:06:125  💛 [MessageSender.m:468 -[MessageSender sendMessage:success:failure:]_block_invoke]: Sending message: OWSSyncContactsMessage, timestamp: 1610300166042.
2021/01/10 17:36:06:126  💛 [TSNetworkManager.m:399 -[TSNetworkManager makeRequestSync:completionQueue:success:failure:]]: Making Non-UD request: { GET: v3/attachments/form/upload }
2021/01/10 17:36:06:151  💛 [JobQueue.swift:214 restartOldJobs()]: marking old `running` IncomingContactSync JobRecords as ready: 0
2021/01/10 17:36:06:178  💛 [JobQueue.swift:214 restartOldJobs()]: marking old `running` IncomingGroupSync JobRecords as ready: 0
2021/01/10 17:36:06:710  💛 [TSNetworkManager.m:417 -[TSNetworkManager makeRequestSync:completionQueue:success:failure:]_block_invoke]: Non-UD request succeeded : { GET: v1/storage/auth }
2021/01/10 17:36:06:712  💛 [TSNetworkManager.m:417 -[TSNetworkManager makeRequestSync:completionQueue:success:failure:]_block_invoke]: Non-UD request succeeded : { GET: v3/attachments/form/upload }
2021/01/10 17:36:06:715  💛 [StorageService.swift:406 storageRequest(withMethod:endpoint:body:)]: Storage request started: get v1/storage/manifest/version/1187
2021/01/10 17:36:06:890  💛 [StorageService.swift:442 storageRequest(withMethod:endpoint:body:)]: Storage request succeeded: get v1/storage/manifest/version/1187
2021/01/10 17:36:06:891  💛 [StorageServiceManager.swift:300 run()]: backup
2021/01/10 17:36:07:005  ❤️ [EarlyMessageManager.swift:58 didReceiveMemoryWarning()]: Dropping all early messages due to memory warning.
2021/01/10 17:36:07:006  ❤️ [EarlyMessageManager.swift:58 didReceiveMemoryWarning()]: Dropping all early messages due to memory warning.
2021/01/10 17:36:07:026  ❤️ [EarlyMessageManager.swift:58 didReceiveMemoryWarning()]: Dropping all early messages due to memory warning.
2021/01/10 17:36:07:343  💛 [Contact.m:106 -[Contact initWithSystemContact:]]: For perf, only considering the first 50 phone numbers for contact with many numbers.

**Comment by ooottttooo** [comment]
Thank you. That solved it for me as well. Life is a little better now 😀

**Comment by dividuum** [comment]
Can also confirm that this seems to be the source of this issue as denying contact access works here too.

I'm surprised why this is the problem: are contacts loaded or stored inefficiently for the share sheet? Even thousands of contacts should fit comfortably into 1-2 MB at most and according to my log from above, the total memory available seems to be 120MB.

The share sheet not working with lots contacts seems to an issue worth investigating more than it has been as it prevents one major user work flow. Is there anything that can be done as someone with no build environment to run the source?

**Comment by dividuum** [comment]
I've never developed for iOS, so my analysis might be totally wrong, but it seems once the share sheet opens:

* We end up [here](https://github.com/signalapp/Signal-iOS/blob/6122856c32276efe8cacf608d39cf971a7844677/SignalShareExtension/ShareViewController.swift#L195) and contacts are loaded
* Calls  [fetchSystemContactsOnceIfAlreadyAuthorized](https://github.com/signalapp/Signal-iOS/blob/112b7c9107a395c3524935bfd9a49e65c2ed3592/SignalMessaging/contacts/OWSContactsManager.m#L147)
* Calls [fetchOnceIfAlreadyAuthorized](https://github.com/signalapp/Signal-iOS/blob/b642a97fec00232640cbb2413c15d2ee4f31acfb/SignalMessaging/contacts/SystemContactsFetcher.swift#L282)
* Calls [updateContacts](https://github.com/signalapp/Signal-iOS/blob/b642a97fec00232640cbb2413c15d2ee4f31acfb/SignalMessaging/contacts/SystemContactsFetcher.swift#L322)
* Calls [fetchContacts](https://github.com/signalapp/Signal-iOS/blob/8cc33b3de17d39b6b827ccc5478764011578a6d0/Signal/src/contact/SystemContactsFetcher.swift#L163)
* It retrieves all contacts using `ABAddressBookCreateWithOptions` and maps them to a `Contact`, the signal internal representation. For that, [buildContact](https://github.com/signalapp/Signal-iOS/blob/8cc33b3de17d39b6b827ccc5478764011578a6d0/Signal/src/contact/SystemContactsFetcher.swift#L174) is called for each contact.
* For extraction of data from the system contact data struct into the internal signal structure, [OWSABRecord](https://github.com/signalapp/Signal-iOS/blob/8cc33b3de17d39b6b827ccc5478764011578a6d0/Signal/src/contact/SystemContactsFetcher.swift#L203) is used.
* Most of the data seem small (just strings, etc..), except for the contact images: It seems those are converted into UIImage objects for each contact [here](https://github.com/signalapp/Signal-iOS/blob/8cc33b3de17d39b6b827ccc5478764011578a6d0/Signal/src/contact/SystemContactsFetcher.swift#L248-L256).

From my (probably poor understanding) it seems that if you have a lot of contacts and a sizable portion of those has pictures assigned, those images are all loaded every time the share sheet opens. So my even poorer fix would be to either scale them down to a reasonable size while doing this import (at least for the share sheet I don't need hi-res images) or lazy load and discard the image data while using them.

**Comment by thealexmack** [comment]
> Can also confirm that this seems to be the source of this issue as denying contact access works here too.
> 
> I'm surprised why this is the problem: are contacts loaded or stored inefficiently for the share sheet? Even thousands of contacts should fit comfortably into 1-2 MB at most and according to my log from above, the total memory available seems to be 120MB.
> 
> The share sheet not working with lots contacts seems to an issue worth investigating more than it has been as it prevents one major user work flow. Is there anything that can be done as someone with no build environment to run the source?

I don’t think it is unique to large numbers of contacts. I’ve been downsizing and am currently at ≈360 yet this issue persists with access to contacts enabled.

**Comment by dividuum** [comment]
> I don’t think it is unique to large numbers of contacts. I’ve been downsizing and am currently at ≈360 yet this issue persists with access to contacts enabled.

Correct. This issue motivated me to clean up my (partly auto-generated) contact list and it's now down to ~150. Still have the issue with access to contacts enabled.

I've also looked into the code again and I think I took a [wrong turn somewhere](https://github.com/signalapp/Signal-iOS/blob/8cc33b3de17d39b6b827ccc5478764011578a6d0/Signal/src/contact/SystemContactsFetcher.swift#L291-L297). The call flow diverts for different iOS version and I think I took the one for ancient versions. The code for current version works a bit different and I didn't have time to look into it yet.

**Comment by bourdoncle** [comment]
Same problem here. New to Signal after dropping WhatsApp. Such a disappointment.

**Comment by dividuum** [comment]
Just tested the new setting "Use system contact photos" in Settings > Appearance. Neither enabling or disabling this setting helps resolve this issue. So it might not be related to system contact photos but something else.

**Comment by Quorum75** [comment]
> Just tested the new setting "Use system contact photos" in Settings > Appearance. Neither enabling or disabling this setting helps resolve this issue. So it might not be related to system contact photos but something else.

Not working for me. At the present time the only work around is to disable the access to Contacts in iOS which is ridiculous. 

Maybe with people switching from WhatsApp to Signal who might complain will force Signal to solve this issue when nothing has been done since 2 years now.

**Comment by thealexmack** [comment]
Just did the most recent update…

My share sheet option is no longer available to signal. I had it saved as a favourite and it is no longer listed as an option to add.

**Comment by ardakara** [comment]
This is now fixed for me as of 5.3 today :)

**Comment by dividuum** [comment]
> This is now fixed for me as of 5.3 today :)

Can confirm. Related commits seem to be https://github.com/signalapp/Signal-iOS/commit/fdaff8918e2e01bf457820f76c6e561648c0f6f9 and https://github.com/signalapp/Signal-iOS/commit/23dc0f021cde8538cd28ae449045174b81294758. Very nice.

**Comment by ardakara** [comment]
Thanks @charlesmchen-signal !
